solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -1357,7 +1357,14 @@ public String register(String coreName, final CoreDescriptor desc, boolean recov
         // we will call register again after zk expiration and on reload
         if (!afterExpiration && !core.isReloaded() && ulog != null && !isTlogReplicaAndNotLeader) {
           // disable recovery in case shard is in construction state (for shard splits)
-          Slice slice = getClusterState().getCollection(collection).getSlice(shardId);
+          DocCollection coll = getClusterState().getCollectionOrNull(collection);
+          if(coll == null) {
+            //unlikely. But if we missed a zk state update we can do a force update
+            // and get again
+            zkStateReader.forceUpdateCollection(collection);
+            coll = getClusterState().getCollection(collection);
+          }
+          Slice slice = coll.getSlice(shardId);
```

Comment:
We may want to check if collection still there

Reply:
`getClusterState().getCollection(collection)`
ensures that the collection is there. if not an Exception is thrown

if `zkStateReader.forceUpdateCollection(collection)` could not already find the collection, it is actually not there